Feat/speedup l1infotree sync

IvanBelyakoff opened the pull request:
Made log processing immediate and hash calculation via a pool of workers. Building tree is done after all logs collected as before. Dedup of BlockByNumber calls and cache headers.

Verification:
- Synced integration8 - OK
- Synced Bali - OK (6h34m -> 5.5m)

IvanBelyakoff committed:
Fixed l1infotree retries on error. Cleanup

IvanBelyakoff committed:
Fixed wrong log processed. Clean up

hexoscott commented on zk/l1infotree/updater.go:
```diff
@@ -103,36 +106,206 @@ func (u *Updater) WarmUp(tx kv.RwTx) (err error) {
+}
+
+func (pool *L1InfoWorkerPool) AddTask(task *L1InfoTask) {
+	pool.taskChan <- task
```
potential to write to a closed channel?  Maybe we could have a status on the worker pool so we can detect that kind of thing?

IvanBelyakoff replied:
I dont think we should do it here. This is a caller's responsibility to make sure that it does not write  to a closed channel and I do make sure I dont do that - in case of syncer is not downloading I stop, wait for workers to finish and return - no way to accidentally write to this channel from callers side.
But you raised a good point so I reviewed another place - requeueing the task in updater could indeed cause panic. Made sure we handle ctx Done channel first to avoid this panic.
Thanx!

hexoscott commented on zk/l1infotree/updater.go:
```diff
@@ -103,36 +106,206 @@ func (u *Updater) WarmUp(tx kv.RwTx) (err error) {
+			time.Sleep(1 * time.Second)
+			go func() {
+				worker.taskChan <- task // requeue the task if there was an error
+			}()
```
Agreed here, we should have some kind of hard cap on how long we wait to re-queue something rather than launching a new go routine without a way to stop it

hexoscott replied:
Also, kind of the task I was looking at, but we will retry indefinitely here which can block the sequencer from producing blocks if the L1 endpoint is down for a long time.

IvanBelyakoff replied:
yeah, I fixed it as I thought it was a valid command. But I was thikning if we should quit with panic on some retry limit or keep retrying forever (because without it there is not much sense of proceeding) until the support team notices something is off and decided to keep retrying.
But thinking on it again, as you mentioned, it might be a valid point we still want to produce blocks, but for how long? And are there any other drawbacks if we have L1 info tree update not finished but we keep producing blocks?

hexoscott replied:
The only drawback is delayed bridging, so we need to shout loud and clear in the logs that this is happening with some clear patter in the log message that we can set up some alerts for 👌

hexoscott commented on zk/l1infotree/updater.go:
```diff
@@ -103,36 +106,206 @@ func (u *Updater) WarmUp(tx kv.RwTx) (err error) {
+	header, err := t.syncer.GetHeader(t.Log.BlockNumber)
+	if err != nil {
+		log.Error("Failed to get header for L1InfoTask", "blockNumber", t.Log.BlockNumber, "error", err)
+		return &L1InfoTaskResult{nil, NewLogKey(t.Log), fmt.Errorf("header not found for block number %d", t.Log.BlockNumber)}
```
let's wrap the original error here, might contain useful details for us debugging later

hexoscott commented on zk/l1infotree/updater.go:
```diff
@@ -103,36 +106,206 @@ func (u *Updater) WarmUp(tx kv.RwTx) (err error) {
+		return &L1InfoTaskResult{nil, NewLogKey(t.Log), fmt.Errorf("header not found for block number %d", t.Log.BlockNumber)}
+	}
+
+	if len(t.Log.Topics) == 3 {
```
Could we invert this if statement and have an early return if it isn't 3?  It would help keep the "line of sight" on the happy path.

hexoscott commented on zk/l1infotree/updater.go:
```diff
@@ -103,36 +106,206 @@ func (u *Updater) WarmUp(tx kv.RwTx) (err error) {
 		}
+	}()
+
+	tree, err := InitialiseL1InfoTree(hermezDb)
```
This can be quite an expensive call to make if there are no updates to process.  The sequencer will run this every 10 seconds or so whilst processing a batch so if we can, could we make this only happen once we know for sure there is at least one log to process?

IvanBelyakoff replied:
oh, allright, wil check that! I did not think it is expensive actually! thanks

hexoscott commented on zk/l1infotree/updater.go:
```diff
@@ -148,82 +321,69 @@ LOOP:
+			update, ok := indexUpdateMap[NewLogKey(l)]
+			if !ok {
+				log.Error(fmt.Sprintf("[%s] Could not find L1 info tree update for log: %v", logPrefix, l))
+				panic(fmt.Sprintf("Could not find L1 info tree update for log: %v", l))
```
Do we need the panic here?  Or is this some seriously broken thing as we MUST have the update in our map at this point in time?

hexoscott reviewed (changes requested):
Looks great, but a few little bits here and there that we might want to consider.

IvanBelyakoff committed:
fix: l1 info tree sync fixes and improvements

Fix potential writing to closed channel
Pass ctx to updated to gracefully shutdown
Add tests for updater

IvanBelyakoff reviewed (commented):
Thanx, @hexoscott for valueable feedback!

IvanBelyakoff committed:
Fix issues with task requeue and completeness.

IvanBelyakoff committed:
Make l1 info channels buffered. Increase batch workers. Cleanup logs

IvanBelyakoff committed:
feat: Improve l1infotree sync speed

IvanBelyakoff committed:
Add forgotten tests.
Fix done channel handling.

IvanBelyakoff committed:
Add L1InfoStage tests
Some channels sync fixes.
Not working again (at least tests are stuck)

IvanBelyakoff committed:
More fixes

IvanBelyakoff committed:
Clean up L1 info tree.
Fix rebase issues.

hexoscott commented on zk/l1infotree/info_tree_l2_rpc_syncer.go:
```diff
@@ -59,7 +59,10 @@ func (s *InfoTreeL2RpcSyncer) RunSyncInfoTree() <-chan []zkTypes.L1InfoTreeUpdat
-		defer close(s.infoTreeChan)
+		defer func() {
+			close(s.infoTreeChan)
+			s.isSyncStarted.Store(true)
```
Should this be false on the `isSyncStarted`?

IvanBelyakoff replied:
correct, nice catch. that change was done during rebasing and I missed it!

hexoscott commented on zk/l1infotree/updater.go:
```diff
@@ -102,36 +111,318 @@ func (u *Updater) WarmUp(tx kv.RwTx) (err error) {
+	}
+
+	if len(t.Log.Topics) != 3 {
+		return &L1InfoTaskResult{nil, NewLogKey(t.Log), nil}
```
Should this be an error state if the log doesn't contain the topics we're expecting?

IvanBelyakoff replied:
Great question. In updater we dont know how many V1 logs we will have. But what we know is how many logs syncer returned when it is done. We use that amount of logs to know when to quit  processig tasks - when the same amount of tasks is finished. I create a task for each log, but I dont want syncer to mess with topics, as I believe its goal is just to fetch logs and headers if needed, but not mess with the topics and their processing.
So here we return an error only if a transient one, then we ll retry. A wrong topic is not an error.

hexoscott commented on zk/syncer/l1_syncer.go:
```diff
@@ -188,25 +193,48 @@ func (s *L1Syncer) RunQueryBlocks(lastCheckedBlock uint64) {
-					s.isDownloading.Store(true)
-					if err := s.queryBlocks(); err != nil {
+					// It should not be checked again in the new cycle, so +1 is added here.
+					startBlock := s.lastCheckedL1Block.Load() + 1
```
This I'm worried about.  In `CheckL2RpcForInfoTreeUpdates` we +1 to the progress of the stage before saving it to the DB.  Lines 739-745.  If we do that on initial sync then the `RunQueryBlocks` also adds +1 here then we can miss a block.  Although we will recognise there is a problem and unwind the tree it could cause us some headaches.  I think we should update the L2 updater to not +1 to the progress in this case.  If we process a leaf twice that is much better than missing one.

IvanBelyakoff replied:
I will check it. Frankly, I did not change that part in terms of ranges, so should be the same it was. because I just moved the range outside the function. it is not a needed change actually, I was just going to change some more stuff with this, but never did it.

hexoscott replied:
Ah, I'm not suggesting you changed something, just that the code as it stands is broken from the L2 syncer point of view, we have the potential to skip a block and miss an event.

hexoscott commented on go.mod:
```diff
@@ -132,6 +132,7 @@ require (
 	github.com/huandu/xstrings v1.4.0 // indirect
 	github.com/munnerz/goautoneg v0.0.0-20191010083416-a7dc8b61c822 // indirect
 	github.com/pion/transport/v3 v3.0.7 // indirect
+	github.com/stretchr/objx v0.5.2 // indirect
```
This seems a little strange on it's own, it's an indirect dependency introduced but with no changes to dependencies?  Maybe some hangover from an earlier iteration?

IvanBelyakoff replied:
ran go mod tidy, it is ok now.

hexoscott commented on zk/syncer/l1_syncer.go:
```diff
@@ -597,3 +622,7 @@ func (s *L1Syncer) QueryForRootLog(to uint64) (*ethTypes.Log, error) {
 	return &logs[0], nil
 }
+
+func (s *L1Syncer) writeDone(numLogs uint64) {
```
Is there another way we could handle this?  There are a few places that take the done channel and when it is triggered we break out of loops.  This is good for an initial sync as we tackle the whole block range in one hit, but beyond that the syncer will continue to run and other processes like the sequencer could receive a false "stop signal" just because of timing of the syncer.

I'm not sure that it's a good signal of being "done" with the task as we write to it many times and it's not a buffered channel so the values could build up over time or the sequencer will read many items from it.  I think we should think about a separate mechanism to signal "we've loaded data for all the available blocks right now", maybe a sync.Cond so components that do care about that can wait, and things like the sequencer can just ignore it.

IvanBelyakoff replied:
I will check that. I dont like it myself, but spent too much time and it is working and the new tests are passing and I thought maybe that is enough.

hexoscott commented on zk/l1infotree/updater.go:
```diff
@@ -102,36 +111,318 @@ func (u *Updater) WarmUp(tx kv.RwTx) (err error) {
+				log.Debug("Retrying L1 info task", "logKey", res.logKey, "error", res.err)
+				// assume a transient error and try again
+				worker.waitGroup.Add(1)
+				time.AfterFunc(1*time.Second, func() {
```
If we fail to re-queue the task because the channel is already at capacity or we waited longer than 10 seconds could we skip a task without requeueing it?  Maybe this needs to be in a loop to keep retrying the requeue?

IvanBelyakoff replied:
the loop will quit on worker.ctx.Done() channel fired, because idleTicker in outer function will fire and will stop the pool by cancelling the context.

hexoscott reviewed (changes requested):
Going to leave the review here.  There is one critical thing around the +1 on the start block that we need to address for sure because it could cause us to miss a leaf.

I like the changes conceptually but I'm worried the code is more complex than it need be, there a few mechanisms for synchronicity from the updater and l1 syncer.  Is there an opportunity to simplify things down using sync.Cond in some places to notify of work completed for example.

hexoscott commented:
Just thinking aloud here... we have the l1 syncer which goes and gets the logs.  Then there is the updater which constructs a worker pool to work through those logs and get the headers... would it be simpler to manage all of that from just the L1 syncer component?  An optional argument to say "also please get me the header for this log" as it syncs?

There would be a lot less co-ordination required this way.

IvanBelyakoff committed:
Refactor L1 info tree synchronization

IvanBelyakoff committed:
Cleaned up

hexoscott commented on zk/syncer/l1_syncer.go:
```diff
@@ -80,9 +84,14 @@ type L1Syncer struct {
 	logsChan         chan []ethTypes.Log
+	logsQueue        []ethTypes.Log
 	logsChanProgress chan string
+	logsChanMutex    sync.Mutex
```
A mutex over a channel is a bit of an antipattern, how come we need this?

IvanBelyakoff replied:
it is used for flushing the queue as it is accessed from both routines. Now I updated the PR, the mutex is used to swap the channels.

hexoscott commented on zk/syncer/l1_syncer.go:
```diff
@@ -148,15 +153,39 @@ func (s *L1Syncer) WaitQueryBlocksToFinish() {
+	// Flush the queue if there are any logs
+	if len(s.logsQueue) > 0 {
+		go func() {
+			s.logsChanMutex.Lock()
```
hmmm, locking the same mutex as in the function start, this could cause a deadlock depending on the timing of the fork/join, the go routine could be started immediately.

IvanBelyakoff replied:
indeed. I updated the PR.thanx

hexoscott reviewed (changes requested):
Leaving the review a little early after spotting some things that we need to address.  Having a mutex over a channel is an antipattern.  To me it suggests that we should perhaps stop using channels and opt for maps/slices instead to keep things simpler.

If we put logs into a map as we download them (protected by a mutex) from the workers with the key being something we can use to identify each log.  Then when the updater asks for them we can just clone that map and have some function to ACK that the log is now processed and the syncer can just drop it from memory then.  It might help with the synchronisation between the two components and allow for better error handling as that is a nightmare with channels.

IvanBelyakoff committed:
Replaced L1Syncer pending logs queue with a channel and removed assignment of nil

IvanBelyakoff committed:
update Updaters activity for big ranges that dont have logs

IvanBelyakoff committed:
Fix task retry.
Removed unneeded getLogs max retry exit - handled by idle timeout.
Add SetFetchHeaders in test

IvanBelyakoff committed:
Add ticker to check for no activity in l1 sync stages

hexoscott reviewed (approved):
Approved 🎆 

Just this CI thing to figure out with the Pectra tests

hexoscott commented:
There is a known problem across all PRs with this pectra type 1 check which looks related to some change in lighthouse, you can merge this once the mandatory tests have passed no problem.
